New CBMC version seems broken in several s2n proofs

<!---
Thank you for reporting a problem and suggesting improvements. Please provide the below information to make sure we can effectively deal with the issue reported. For the most precise version information, see the first line of console output or run with --version. Please attach or include example code that allows us to reproduce the problem.
--->

CBMC version: 5.19.0
Operating system: macOS High Sierra, 10.13.6
Exact command line resulting in the issue: `make report` in any proof harnesses from [s2n codebase](https://github.com/awslabs/s2n) under `s2n/tests/cbmc/proofs/` folder.
What behaviour did you expect: Run `goto-cc`, `goto-instrument` and `cbmc` without errors.
What happened instead: Invariant violation on `goto-cc`.

```shell
Invariant check failed
File: /tmp/cbmc-20201130-38442-24cwlx/src/util/mathematical_expr.cpp:22 function: function_application_exprt
Condition: domain.size() == arguments().size()
Reason: Precondition
Backtrace:
0   goto-cc                             0x000000010c077e4a _Z15print_backtraceRNSt3__113basic_ostreamIcNS_11char_traitsIcEEEE + 74
1   goto-cc                             0x000000010c078354 _Z13get_backtracev + 174
2   goto-cc                             0x000000010be7109c _Z29invariant_violated_structuredI17invariant_failedtJRKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEEENS1_9enable_ifIXsr3std10is_base_ofIS0_T_EE5valueEvE4typeES9_S9_iS9_DpOT0_ + 44
3   goto-cc                             0x000000010bea04a9 _Z25invariant_violated_stringRKNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEES7_iS7_S7_ + 9
4   goto-cc                             0x000000010c0841ee _ZN26function_application_exprtC2ERK5exprtNSt3__16vectorIS0_NS3_9allocatorIS0_EEEE + 750
5   goto-cc                             0x000000010bf34247 _ZN13goto_convertt23do_function_call_symbolERK5exprtRK12symbol_exprtRKNSt3__16vectorIS0_NS6_9allocatorIS0_EEEER13goto_programt + 4689
6   goto-cc                             0x000000010bf543e0 _ZN13goto_convertt16do_function_callERK5exprtS2_RKNSt3__16vectorIS0_NS3_9allocatorIS0_EEEER13goto_programtRK8dstringt + 408
7   goto-cc                             0x000000010bf5b69b _ZN13goto_convertt20remove_function_callER31side_effect_expr_function_calltR13goto_programtRK8dstringtb + 771
8   goto-cc                             0x000000010bf3e59c _ZN13goto_convertt10clean_exprER5exprtR13goto_programtRK8dstringtb + 3016
9   goto-cc                             0x000000010bf3e59c _ZN13goto_convertt10clean_exprER5exprtR13goto_programtRK8dstringtb + 3016
10  goto-cc                             0x000000010bf3e59c _ZN13goto_convertt10clean_exprER5exprtR13goto_programtRK8dstringtb + 3016
11  goto-cc                             0x000000010bf44d8e _ZN13goto_convertt18convert_expressionERK16code_expressiontR13goto_programtRK8dstringt + 578
12  goto-cc                             0x000000010bf41da1 _ZN13goto_convertt7convertERK5codetR13goto_programtRK8dstringt + 185
13  goto-cc                             0x000000010bf44407 _ZN13goto_convertt13convert_blockERK11code_blocktR13goto_programtRK8dstringt + 125
14  goto-cc                             0x000000010bf41d40 _ZN13goto_convertt7convertERK5codetR13goto_programtRK8dstringt + 88
15  goto-cc                             0x000000010bf4a761 _ZN13goto_convertt18convert_ifthenelseERK16code_ifthenelsetR13goto_programtRK8dstringt + 657
16  goto-cc                             0x000000010bf42076 _ZN13goto_convertt7convertERK5codetR13goto_programtRK8dstringt + 910
17  goto-cc                             0x000000010bf44407 _ZN13goto_convertt13convert_blockERK11code_blocktR13goto_programtRK8dstringt + 125
18  goto-cc                             0x000000010bf41d40 _ZN13goto_convertt7convertERK5codetR13goto_programtRK8dstringt + 88
19  goto-cc                             0x000000010bf418ce _ZN13goto_convertt16goto_convert_recERK5codetR13goto_programtRK8dstringt + 34
20  goto-cc                             0x000000010bf565b8 _ZN23goto_convert_functionst16convert_functionERK8dstringtR14goto_functiont + 1318
21  goto-cc                             0x000000010be79673 _ZN8compilet15convert_symbolsER15goto_functionst + 483
22  goto-cc                             0x000000010be77c44 _ZN8compilet4linkEv + 92
23  goto-cc                             0x000000010be76601 _ZN8compilet4doitEv + 577
24  goto-cc                             0x000000010be8f467 _ZN9gcc_modet4doitEv + 9121
25  goto-cc                             0x000000010be94aca _ZN13goto_cc_modet4mainEiPPKc + 56
26  goto-cc                             0x000000010be6e5b1 main + 2804
27  libdyld.dylib                       0x00007fff62101015 start + 1
28  ???                                 0x0000000000000018 0x0 + 24
--- end invariant violation report ---
```

Reproduced for the `s2n_array_new` proof, but the problem should be fixed in s2n. Compiling with `-Wall` added to compile options:
```
/home/ub-backup/tautschn/s2n.git/tests/cbmc/sources/cbmc_utils.c: In function 'uninterpreted_compare':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/sources/cbmc_utils.c:132:1: warning: function '__CPROVER_uninterpreted_hasher' is not declared
     if (rval == 0) { __CPROVER_assume(__CPROVER_uninterpreted_hasher(a) == __CPROVER_uninterpreted_hasher(b)); }
/home/ub-backup/tautschn/s2n.git/tests/cbmc/sources/cbmc_utils.c: In function 'nondet_s2n_mem_init':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/sources/cbmc_utils.c:187:1: warning: function 's2n_mem_init' is not declared
     if (nondet_bool()) { s2n_mem_init(); }
/home/ub-backup/tautschn/s2n.git/tests/cbmc/sources/make_common_datastructures.c: In function 'nondet_comparator':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/sources/make_common_datastructures.c:93:1: warning: function 'assert' is not declared
     assert(a != NULL);
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/mlock.c: In function 'mlock':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/mlock.c:23:1: warning: function 'assert' is not declared
     assert(S2N_MEM_IS_WRITABLE(addr, len));
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/munlock.c: In function 'munlock':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/munlock.c:23:1: warning: function 'assert' is not declared
     assert(S2N_MEM_IS_WRITABLE(addr, len));
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/posix_memalign_override.c: In function 'posix_memalign':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/posix_memalign_override.c:36:1: warning: function 'malloc' is not declared
     void *tmp = malloc(size);
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/posix_memalign_override.c: In function 'posix_memalign':
/home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/posix_memalign_override.c:36:1: warning: warning: ignoring conflicting implicit function declaration "malloc"
old definition in module make_common_datastructures file /usr/include/stdlib.h line 539
void * (size_t)
new definition in module posix_memalign_override file /home/ub-backup/tautschn/s2n.git/tests/cbmc/stubs/posix_memalign_override.c line 36 function posix_memalign
```
The main problem is the first warning: `__CPROVER_uninterpreted_hasher` is not declared, and thus we don't know the correct type of the (uninterpreted) function. I'll try to come up with a better error message, but this will just make your failure easier to understand. The fix, as said, needs to go into the s2n repository.

--- a/regression/ansi-c/undeclared_function/undeclared_uninterpreted.desc
+++ b/regression/ansi-c/undeclared_function/undeclared_uninterpreted.desc
@@ -1,0 +1,11 @@
+CORE
+uninterpreted.c
+
+missing type information required to construct call to uninterpreted function
+^EXIT=70$
+^SIGNAL=0$
+--
+^warning: ignoring
+^Invariant check failed$
+--
+Test to confirm that an actionable error message is provided.

--- a/regression/ansi-c/undeclared_function/uninterpreted.c
+++ b/regression/ansi-c/undeclared_function/uninterpreted.c
@@ -1,0 +1,10 @@
+#include <assert.h>
+
+int main()
+{
+  int x;
+  assert(__CPROVER_uninterpreted_fn(x) == __CPROVER_uninterpreted_fn(x));
+  return 0;
+}
+
+int __CPROVER_uninterpreted_fn(int x);

--- a/src/goto-programs/builtin_functions.cpp
+++ b/src/goto-programs/builtin_functions.cpp
@@ -892,6 +892,15 @@ void goto_convertt::do_function_call_symbol(
     if(lhs.is_nil())
       return;
 
+    if(function.type().get_bool(ID_C_incomplete))
+    {
+      error().source_location = function.find_source_location();
+      error() << "'" << identifier << "' is not declared, "
+              << "missing type information required to construct call to "
+              << "uninterpreted function" << eom;
+      throw 0;
+    }
+
     const code_typet &function_call_type = to_code_type(function.type());
     mathematical_function_typet::domaint domain;
     for(const auto &parameter : function_call_type.parameters())

